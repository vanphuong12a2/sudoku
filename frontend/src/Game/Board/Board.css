.main-board {
    margin:1em;
    max-width: 450px;
    max-height: 450px;
    width: 100%;
    position: relative;
}

.main-board td {
    border:1px solid;
    padding: 0;
}
.main-board td:first-child {
    border-left:solid;
}
.main-board td:nth-child(3n) {
    border-right:solid ;
}
.main-board tr:first-child {
    border-top:solid;
}
.main-board tr:nth-child(3n) td {
    border-bottom:solid ;
}

.main-board caption {
    text-align: left;
}

.main-board input,
.main-board input:hover,
#root .main-board input:focus {
    width: 50px;
    height: 50px;
    text-align:center;
    border: 0;
    margin: 0;
    font-size: 1.7rem;
}

.main-board input.writable {
    color: green;
}

.main-board input.readonly {
    cursor: pointer;
    color: black;
}

.main-board input.invalid {
    background: pink;
    border: 0;
    box-shadow: none;
}

.main-board i {
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    font-size: 3em;
}

