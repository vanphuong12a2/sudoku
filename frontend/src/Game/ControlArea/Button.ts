export default class Button {
    public name: string;
    public onClickHandler: () => void;

    constructor(name: string, onClickHandler: () => void) {
        this.name = name;
        this.onClickHandler = onClickHandler;
    }
}