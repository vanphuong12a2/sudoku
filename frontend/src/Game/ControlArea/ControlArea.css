.control-area {
    margin: 1em;
}