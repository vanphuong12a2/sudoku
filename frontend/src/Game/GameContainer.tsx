import React from 'react';
import Game from './Game';
import {generateBoard, generateEmptyBoard} from '../common/boardFunctions';

interface Props {
}

interface State {
    currentBoard: number[][]
}

class GameContainer extends React.Component<{}, State> {

    constructor(props: Props) {
        super(props);
        this.state = {currentBoard: generateBoard()}
    }

    render() {
        return (
            <Game
                boardData={this.state.currentBoard}
                newGameOnClickHandler={this.newGameOnClickHandler}
                clearBoardOnClickHandler={this.clearBoardOnClickHandler}
            />);
    }

    public clearBoardOnClickHandler = () => {
        this.setState({currentBoard: generateEmptyBoard()});
    };

    public newGameOnClickHandler = () => {
        this.setState({currentBoard: generateBoard()});
    };
}

export default GameContainer;