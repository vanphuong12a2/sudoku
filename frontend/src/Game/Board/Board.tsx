import React from 'react';
import Cell, {EMPTY_CELL} from '../Cell/Cell';
import './Board.css';

interface Props {
    loading: boolean
    boardData: number[][]
    currentBoard: number[][]
    onCellChange: (rowIndex: number, cellIndex: number) => (newValue: number) => void
}

const Board = (props: Props) => {

    const boardTable = <table>
        <tbody>
        {props.currentBoard.map((row, rowIndex) => {
            return (
                <tr key={rowIndex}>
                    {
                        props.currentBoard[rowIndex].map((cellData, cellIndex) => {
                            return (
                                <Cell
                                    key={cellIndex}
                                    cellData={cellData}
                                    readOnly={props.boardData[rowIndex][cellIndex] !== EMPTY_CELL}
                                    onCellChange={props.onCellChange(rowIndex, cellIndex)}
                                />
                            );
                        })
                    }
                </tr>);
        })}
        </tbody>
    </table>;

    const loadingSpinner = <i className='fa fa-spinner fa-spin'/>;

    return (
        <div className='main-board'>
            {props.loading ? loadingSpinner : boardTable}
        </div>
    );
};

export default Board;