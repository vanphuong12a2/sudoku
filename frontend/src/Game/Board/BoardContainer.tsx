import React from 'react';
import Board from './Board';

interface Props {
    boardData: (number)[][]
}

interface State {
    currentBoard: (number)[][]
}

class BoardContainer extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.state = {currentBoard: this.copyBoard(this.props.boardData)};
    }

    private copyBoard = (board: number[][]) => {
        return board.map(row => row.slice());
    };

    componentWillReceiveProps(nextProps: Props) {
        if (nextProps.boardData !== this.props.boardData) {
            this.setState({currentBoard: this.copyBoard(nextProps.boardData)});
        }
    }

    render() {
        return (
            <Board
                boardData={this.props.boardData}
                currentBoard={this.state.currentBoard}
                onCellChange={this.onCellChange}
            />
        )
    }

    public onCellChange = (rowIndex: number, cellIndex: number) => (newValue: number) => {
        this.state.currentBoard[rowIndex][cellIndex] = newValue;
        this.setState({currentBoard: this.state.currentBoard});
    }
}

export default BoardContainer;