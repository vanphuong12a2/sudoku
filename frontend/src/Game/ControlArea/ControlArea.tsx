import React from 'react';
import './ControlArea.css'
import Button from './Button';

const ControlFieldSet = (props: { legend: string, buttons: Button[] }) => {
    return (
        <fieldset>
            <legend>{props.legend}</legend>
            {
                props.buttons.map((button, buttonIndex) => {
                    return <ButtonComponent key={buttonIndex} button={button}/>
                })
            }
        </fieldset>
    );
};

const ButtonComponent = (props: { button: Button }) => {
    return (
        <a className="waves-effect waves-light btn-small" onClick={props.button.onClickHandler}>
            {props.button.name}
        </a>
    );
};

interface Props {
    newGameOnClickHandler: () => void
    clearBoardOnClickHandler: () => void
}

const ControlArea = (props: Props) => {
    let dummyHandler = () => {;};
    const gameButtons = [
        new Button('New game', props.newGameOnClickHandler),
        new Button('Show a hint', dummyHandler)
    ];

    const solverButtons = [
        new Button('Clear', props.clearBoardOnClickHandler),
        new Button('Validate', dummyHandler),
        new Button('Show solution', dummyHandler)
    ];
    return (
        <div className='control-area'>
            <ControlFieldSet legend='Game' buttons={gameButtons}/>
            <ControlFieldSet legend='Solver' buttons={solverButtons}/>
        </div>
    )
};

export default ControlArea;