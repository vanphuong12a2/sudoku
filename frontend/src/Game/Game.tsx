import React from 'react';
import ControlArea from './ControlArea/ControlArea';
import './Game.css'
import Board from './Board/Board';

interface Props {
    boardData: number[][]
    currentBoard: number[][]
    newGameOnClickHandler: () => void
    refreshGameOnClickHandler: () => void
    clearBoardOnClickHandler: () => void
    onCellChange: (rowIndex: number, cellIndex: number) => (newValue: number) => void
}

const Game = (props: Props) => {
    return (
        <div className='game-container'>
            <Board
                boardData={props.boardData}
                currentBoard={props.currentBoard}
                onCellChange={props.onCellChange}
            />
            <ControlArea
                newGameOnClickHandler={props.newGameOnClickHandler}
                refreshGameOnClickHandler={props.refreshGameOnClickHandler}
                clearBoardOnClickHandler={props.clearBoardOnClickHandler}
            />
        </div>
    );
};

export default Game;