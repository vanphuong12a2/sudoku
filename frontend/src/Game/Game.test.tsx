import React from 'react';
import {create} from 'react-test-renderer';
import Game from './Game';
import {sampleBoardData} from '../common/testData';
import {shallow} from 'enzyme';
import Board from './Board/Board';
import ControlArea from './ControlArea/ControlArea';

describe('<Game />', () => {

    describe('Snapshots', () => {

        it('renders game area correctly', () => {
            const board = create(<Game
                boardData={sampleBoardData()}
                currentBoard={sampleBoardData()}
                onCellChange={jest.fn()}
                newGameOnClickHandler={jest.fn()}
                refreshGameOnClickHandler={jest.fn()}
                clearBoardOnClickHandler={jest.fn()}
            />);
            expect(board.toJSON()).toMatchSnapshot();
        });
    });

    it('should send correct props to control area', function () {

        let boardData = sampleBoardData();
        let onCellChange = jest.fn();
        let newGameOnClickHandler = jest.fn();
        let refreshGameOnClickHandler = jest.fn();
        let clearBoardOnClickHandler = jest.fn();

        const component = shallow(<Game
            boardData={boardData}
            currentBoard={boardData}
            onCellChange={onCellChange}
            newGameOnClickHandler={newGameOnClickHandler}
            refreshGameOnClickHandler={refreshGameOnClickHandler}
            clearBoardOnClickHandler={clearBoardOnClickHandler}
        />);

        expect(component.find(Board).prop('boardData')).toEqual(boardData);
        expect(component.find(Board).prop('currentBoard')).toEqual(boardData);
        expect(component.find(Board).prop('onCellChange')).toEqual(onCellChange);
        expect(component.find(ControlArea).prop('newGameOnClickHandler')).toEqual(newGameOnClickHandler);
        expect(component.find(ControlArea).prop('refreshGameOnClickHandler')).toEqual(refreshGameOnClickHandler);
        expect(component.find(ControlArea).prop('clearBoardOnClickHandler')).toEqual(clearBoardOnClickHandler);
    });
});