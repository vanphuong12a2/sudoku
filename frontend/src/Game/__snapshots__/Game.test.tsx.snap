// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> Snapshots renders game area correctly 1`] = `
<div
  className="game-container"
>
  <table
    className="main-board"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={2}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={5}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={4}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={1}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={8}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={3}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={7}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={4}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={1}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={8}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={4}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={7}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={5}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={6}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={2}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={8}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={1}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={4}
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={6}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={8}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={7}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={2}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={1}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={8}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={5}
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={9}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={5}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={3}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={2}
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={5}
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={8}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="readonly"
            maxLength={1}
            onChange={[Function]}
            readOnly={true}
            value={7}
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
        <td>
          <input
            className="writable"
            maxLength={1}
            onChange={[Function]}
            readOnly={false}
            value=""
          />
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="control-area"
  >
    <fieldset>
      <legend>
        Game
      </legend>
      <a
        className="waves-effect waves-light btn-small"
        onClick={[MockFunction]}
      >
        Refresh
      </a>
      <a
        className="waves-effect waves-light btn-small"
        onClick={[MockFunction]}
      >
        New game
      </a>
    </fieldset>
    <fieldset>
      <legend>
        Solver
      </legend>
      <a
        className="waves-effect waves-light btn-small"
        onClick={[MockFunction]}
      >
        Clear all
      </a>
      <a
        className="waves-effect waves-light btn-small"
        onClick={[Function]}
      >
        Validate
      </a>
      <a
        className="waves-effect waves-light btn-small"
        onClick={[Function]}
      >
        Show solution
      </a>
    </fieldset>
  </div>
</div>
`;
